services/src/main/java/io/grpc/services/BinaryLog.java

```diff
@@ -46,33 +49,124 @@
   private static final int IP_PORT_LOWER_MASK = 0xff;
 
-  private final int maxHeaderBytes;
-  private final int maxMessageBytes;
+  private final BinaryLogSink sink;
+  @VisibleForTesting
+  final int maxHeaderBytes;
+  @VisibleForTesting
+  final int maxMessageBytes;
 
   @VisibleForTesting
-  BinaryLog(int maxHeaderBytes, int maxMessageBytes) {
+  BinaryLog(BinaryLogSink sink, int maxHeaderBytes, int maxMessageBytes) {
+    this.sink = sink;
     this.maxHeaderBytes = maxHeaderBytes;
     this.maxMessageBytes = maxMessageBytes;
   }
 
+  /**
+   * Logs the sending of initial metadata. This method logs the appropriate number of bytes
+   * as determined by the binary logging configuration.
+   */
+  public void logSendInitialMetadata(
+      Metadata metadata, boolean isServer, byte[] callId, SocketAddress peerSocket) {
+    GrpcLogEntry entry = GrpcLogEntry
+        .newBuilder()
+        .setType(Type.SEND_INITIAL_METADATA)
+        .setLogger(isServer ? GrpcLogEntry.Logger.SERVER : GrpcLogEntry.Logger.CLIENT)
+        .setCallId(callIdToProto(callId))
+        .setPeer(socketToProto(peerSocket))
+        .setMetadata(metadataToProto(metadata, maxHeaderBytes))
+        .build();
+    sink.write(entry);
+  }
+
+  /**
+   * Logs the receiving of initial metadata. This method logs the appropriate number of bytes
+   * as determined by the binary logging configuration.
+   */
+  public void logRecvInitialMetadata(
+      Metadata metadata, boolean isServer, byte[] callId, SocketAddress peerSocket) {
+    GrpcLogEntry entry = GrpcLogEntry
+        .newBuilder()
+        .setType(Type.RECV_INITIAL_METADATA)
+        .setLogger(isServer ? GrpcLogEntry.Logger.SERVER : GrpcLogEntry.Logger.CLIENT)
+        .setCallId(callIdToProto(callId))
+        .setPeer(socketToProto(peerSocket))
+        .setMetadata(metadataToProto(metadata, maxHeaderBytes))
+        .build();
+    sink.write(entry);
+  }
+
+  /**
+   * Logs the trailing metadata. This method logs the appropriate number of bytes
+   * as determined by the binary logging configuration.
+   */
+  public void logTrailingMetadata(Metadata metadata, boolean isServer, byte[] callId) {
+    GrpcLogEntry entry = GrpcLogEntry
+        .newBuilder()
+        .setType(isServer ? Type.SEND_TRAILING_METADATA : Type.RECV_TRAILING_METADATA)
+        .setLogger(isServer ? GrpcLogEntry.Logger.SERVER : GrpcLogEntry.Logger.CLIENT)
+        .setCallId(callIdToProto(callId))
+        .setMetadata(metadataToProto(metadata, maxHeaderBytes))
+        .build();
+    sink.write(entry);
+  }
+
+  /**
+   * Logs the outbound message. This method logs the appropriate number of bytes from
+   * {@code message}, and returns an {@link InputStream} that contains the original message.
+   * The number of bytes logged is determined by the binary logging configuration.
+   */
+  public void logOutboundMessage(
+      byte[] message, boolean compressed, boolean isServer, byte[] callId) {
```

Comment:
It can, although I don't think it's going to make the code much simpler at the moment. Is this to make the signatures match up with some proposed future changes on the Marshaller API?

I also see a UnsafeByteOperations.unsfeWrap() that allows protobuf to avoid making another copy of the contents. It seems appropriate for this use case, since the underlying byte[]/ByteBuffer is only shared between the binary logger and grpc marshaller, and neither will modify the bytes.

Reply:
The change is for performance, not so much for API convenience.